bcda/api/requests.go

```diff
@@ -0,0 +1,292 @@
+		responseutils.WriteError(oo, w, http.StatusInternalServerError)
+		return
+	}
+	newJob.TransactionTime = b.Meta.LastUpdated
+
+	// Decode the _since parameter (if it exists) so it can be persisted in job args
+	var decodedSince string
+	params, ok := r.URL.Query()["_since"]
+	if ok {
+		decodedSince, _ = url.QueryUnescape(params[0])
+	}
+
+	var enqueueJobs []*que.Job
+	enqueueJobs, err = newJob.GetEnqueJobs(resourceTypes, decodedSince, retrieveNewBeneHistData)
+	if err != nil {
+		log.Error(err)
+		oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Processing, "")
+		responseutils.WriteError(oo, w, http.StatusInternalServerError)
+		return
+	}
+	newJob.JobCount = len(enqueueJobs)
+
+	// We've now computed all of the fields necessary to populate a fully defined job
+	if err = tx.Save(&newJob).Error; err != nil {
+		log.Error(err.Error())
+		oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.DbErr, "")
+		responseutils.WriteError(oo, w, http.StatusInternalServerError)
+		return
+	}
+
+	// Since we're enqueuing these queuejobs BEFORE we've created the actual job, we may encounter a transient
+	// error where the job does not exist. Since queuejobs are retried, the transient error will be resolved
+	// once we finish inserting the job.
+	for _, j := range enqueueJobs {
+		if err = qc.Enqueue(j); err != nil {
+			log.Error(err)
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Processing, "")
+			responseutils.WriteError(oo, w, http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func Check429(jobs []models.Job, types []string, w http.ResponseWriter) ([]string, bool) {
+	var unworkedTypes []string
+
+	for _, t := range types {
+		worked := false
+		for _, job := range jobs {
+			req, err := url.Parse(job.RequestURL)
+			if err != nil {
+				log.Error(err)
+				oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Processing, "")
+				responseutils.WriteError(oo, w, http.StatusInternalServerError)
+			}
+
+			if requestedTypes, ok := req.Query()["_type"]; ok {
+				// if this type is being worked no need to keep looking, break out and go to the next type.
+				if strings.Contains(requestedTypes[0], t) && (job.Status == "Pending" || job.Status == "In Progress") && (job.CreatedAt.Add(GetJobTimeout()).After(time.Now())) {
+					worked = true
+					break
+				}
+			} else {
+				// check to see if the export all is still being worked
+				if (job.Status == "Pending" || job.Status == "In Progress") && (job.CreatedAt.Add(GetJobTimeout()).After(time.Now())) {
+					return nil, false
+				}
+			}
+		}
+		if !worked {
+			unworkedTypes = append(unworkedTypes, t)
+		}
+	}
+	if len(unworkedTypes) == 0 {
+		return nil, false
+	} else {
+		return unworkedTypes, true
+	}
+}
+
+func ValidateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
```

Comment:
I'm wondering about function visibility here... should this be an exported function?